### PR TITLE
DataType support for non-strongly typed languages

### PR DESCRIPTION
**Pull request by mathewc**
Notes:
======
- adds support for allowing scripts to declare the **input type** to bind to (addresses https://github.com/Azure/azure-webjobs-sdk-script/issues/197, https://github.com/Azure/azure-webjobs-sdk-script/issues/307)
- the intent is to expose this data type in the portal for Node input bindings (that's the only case its needed currently)
- we can add more to data types in the future, but for now, string/binary covers all the bases. The base bindings don't support int, etc. anyways. Strings can handle all of those cases easily.
- cleaning up the binding pipeline, removing unecessary Stream conversions, etc.
- Fixed "Multiple custom attributes of the same type found" issue with C# functions that have compilation errors
- also addresses part of https://github.com/Azure/azure-webjobs-sdk-script/issues/17

### CODE REVIEW

File: src/WebJobs.Script/Description/DotNet/DotNetFunctionDescriptionProvider.cs [hunk 1]
```diff
@@ -174,6 +174,7 @@ protected override IFunctionInvoker CreateFunctionInvoker(string scriptFilePath,
 
             // We were unable to compile the function to get its signature,
             // setup the descriptor with the default parameters
+            methodAttributes.Clear();
```
Comment (mathewc) [inline src/WebJobs.Script/Description/DotNet/DotNetFunctionDescriptionProvider.cs]: This fixes the "multiple attributes" issue we've been seeing. Above we were adding a [NoAutomaticTrigger] attribute to the methodAttribute collection as part of compilation, but when that fails, the collection still contains that attribute, and the following method adds another, causing the core SDK to throw.

File: src/WebJobs.Script/Description/FunctionDescriptorProvider.cs [hunk 2]
```diff
@@ -134,16 +134,19 @@ protected virtual ParameterDescriptor CreateTriggerParameter(BindingMetadata tri
             switch (triggerMetadata.Type)
             {
                 case BindingType.QueueTrigger:
-                    triggerParameter = ParseQueueTrigger((QueueBindingMetadata)triggerMetadata, parameterType ?? typeof(string));
+                    parameterType = DefaultParameterType(parameterType, triggerMetadata.DataType);
```
Comment (mathewc) [inline src/WebJobs.Script/Description/FunctionDescriptorProvider.cs]: These changes allow the function do declare a byte[] etc. trigger value, so their scripts receive a byte[]. Note that for the bindings that are binding to Stream we don't need to do anything here - the conversion happens later when we read the value from the stream.

File: src/WebJobs.Script/Description/Node/NodeFunctionInvoker.cs [hunk 3]
```diff
@@ -392,39 +366,63 @@ protected override void OnScriptFileChanged(object sender, FileSystemEventArgs e
...
+                    {
+                        body = request.Content.ReadAsStringAsync().Result;
+                        if (TryDeserializeJson((string)body, out jsonObject))
+                        {
+                            // if the content - type of the request is json, deserialize into an object
+                            rawBody = (string)body;
+                            body = jsonObject;
+                        }
+                    }
+                    else if (contentType.MediaType == "application/octet-stream")
```
Comment (mathewc) [inline src/WebJobs.Script/Description/Node/NodeFunctionInvoker.cs]: Still thinking more on the right model for this. In the tests, I've added an http trigger test that specifies dataType binary so the script receives a byte[], however I'm wondering if I should key that off of the dataType only, or content type as I'm doing here.

File: src/WebJobs.Script/Host/ScriptHost.cs [hunk 4]
```diff
@@ -654,7 +654,9 @@ private static void ApplyBindingConfiguration(Collection<FunctionDescriptor> fun
 
         private void HandleHostError(Microsoft.Azure.WebJobs.Extensions.TraceFilter traceFilter)
         {
-            foreach (TraceEvent traceEvent in traceFilter.Events)
+            var events = traceFilter.Events.ToArray();
```
Comment (mathewc) [inline src/WebJobs.Script/Host/ScriptHost.cs]: Was seeing some "collection modified while enumerating" exceptions, so caching this

File: test/WebJobs.Script.Tests/TestScripts/Node/BlobTriggerToBlob/function.json [hunk 5]
```diff
@@ -0,0 +1,16 @@
+﻿{
+  "bindings": [
+    {
+      "type": "blobTrigger",
+      "direction": "in",
+      "dataType": "binary",
```
Comment (mathewc) [inline test/WebJobs.Script.Tests/TestScripts/Node/BlobTriggerToBlob/function.json]: This is the new capability - I declare that I expect binary data, so my Node script gets a Buffer, whereas before we were always converting to string, which doesn't work for things like images, etc.

File: test/WebJobs.Script.Tests/TestScripts/Node/QueueTriggerByteArray/function.json [hunk 6]
```diff
@@ -0,0 +1,16 @@
+﻿{
+  "bindings": [
+    {
+        "type": "queueTrigger",
+        "direction": "in",
+        "dataType": "binary",
```
Comment (mathewc) [inline test/WebJobs.Script.Tests/TestScripts/Node/QueueTriggerByteArray/function.json]: Can also enqueue/dequeue binary messages

File: src/WebJobs.Script/Description/Binding/DataType.cs [hunk 7]
```diff
@@ -0,0 +1,15 @@
...
+namespace Microsoft.Azure.WebJobs.Script.Description
+{
+    /// <summary>
+    /// Enumeration of binding value data types.
+    /// </summary>
+    public enum DataType
+    {
+        String,
+        Binary,
+        Stream
```
Comment (mathewc) [inline src/WebJobs.Script/Description/Binding/DataType.cs]: Stream is only intended to be used internally currently. However, we can investigate adding Streaming support to Node (what you really want for image processing), but that will be a bit of work (we'll need to define an interop between .NET Stream and Node - Edge.js doesn't support Streams currently).

File: src/WebJobs.Script/Description/DotNet/DotNetFunctionDescriptionProvider.cs (hunk 1 above)
Comment (fabiocav) [inline src/WebJobs.Script/Description/DotNet/DotNetFunctionDescriptionProvider.cs]: 🎉

**Comment by mathewc** [comment]
In the latest code I pushed I made a bunch of changes to fix the test flakiness we've been seeing. I'm getting good runs now, whereas before I was getting lots of failures. Specifically I've:

* removed use of ScriptHostManager from the end to end tests - we only need ScriptHost. The long running manager was causing issues
* removed the nuget restore the Mobile/NH tests were doing, and replaced with direct assembly references
* created more test isolation by using unique host IDs for each e2e test suite. Also using unique queues/blobs/etc. per test suite

File: test/WebJobs.Script.Tests/Properties/AssemblyInfo.cs [hunk 8]
```diff
@@ -39,4 +39,4 @@
 [assembly: AssemblyVersion("1.0.0.0")]
 [assembly: AssemblyFileVersion("1.0.0.0")]
 
-[assembly: CollectionBehavior(CollectionBehavior.CollectionPerAssembly)]
+[assembly: CollectionBehavior(CollectionBehavior.CollectionPerAssembly, DisableTestParallelization = true)]
```
Comment (fabiocav) [inline test/WebJobs.Script.Tests/Properties/AssemblyInfo.cs]: Did you find issues with parallelization?

**Comment by fabiocav** [comment]
This is getting quite large to review, I looked over the changes and added a couple of minor comments, but it looks good.
About the failing tests, I still see the Node BlobTriggerToBlobTest failing :( Did you find what was wrong with it?
For the other two that are failing, are we just missing the appropriate containers?

File: test/WebJobs.Script.Tests/Properties/AssemblyInfo.cs (hunk 8 above)
Comment (mathewc) [inline test/WebJobs.Script.Tests/Properties/AssemblyInfo.cs]: No, I was grasping at straws. I already have parallelization turned off, via CollectionPerAssembly (there is only one assembly). I could probably revert this.

**Comment by mathewc** [comment]
@fabiocav Yes this PR has gotten large. There is still more cleanup/simplification I want to do, but I think I'll just get this work in first and do a follow up. I want to add a bunch more tests, and continue the test stabilization work.

**Comment by simeyla** [comment]
@mathewc Is this ever going to make it into the Portal UI?
Took me way too long today to eventually find this and get my blobs to not be all mangled.

**Pull request merged**